### ISSUE TITLE
Broken setup after upgrade

### ISSUE THREAD

**Issue by dimm0** [projectcalico/calico#5309]
I upgraded calico 16.5->21.1, and it seemed fine except 2 nodes which needed a reboot and worked fine after.
Then I upgraded k8s 1.21 -> 1.22. The upgrade went fine, but I noticed old coredns pods were not ready, with error:

```
[ERROR] plugin/errors: 2 2438761344240128283.14968686270740531. HINFO: read udp 10.244.6.130:40715->8.8.8.8:53: i/o timeout
[ERROR] plugin/errors: 2 2438761344240128283.14968686270740531. HINFO: read udp 10.244.6.130:53575->8.8.8.8:53: i/o timeout
[ERROR] plugin/errors: 2 2438761344240128283.14968686270740531. HINFO: read udp 10.244.6.130:40603->8.8.8.8:53: i/o timeout
```

And new ones were starting with the same error. I'm not sure when it broke, after the calico upgrade or kubernetes one. At that point there was some connectivity still in the cluster, but seemed like no UDP traffic was allowed.

Now there's no connectivity between pods even on the same node. Can access pods from the host. Multiple related errors in the cluster, don't even know where to start.

What logs can I provide?

Typha is logging this:

```
2021-12-18 03:28:47.852 [INFO][6] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/kubernetesendpointslices" error=ListOptions.meta.k8s.io "" is invalid: [resourceVersionMatch: Forbidden: resourceVersionMatch is f
2021-12-18 03:28:47.871 [INFO][6] watchercache.go 175: Full resync is required ListRoot="/calico/resources/v3/projectcalico.org/workloadendpoints"
2021-12-18 03:28:48.139 [INFO][6] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/workloadendpoints" error=ListOptions.meta.k8s.io "" is invalid: [resourceVersionMatch: Forbidden: resourceVersionMatch is forbidde
2021-12-18 03:28:48.445 [INFO][6] watchercache.go 175: Full resync is required ListRoot="/calico/resources/v3/projectcalico.org/profiles"
2021-12-18 03:28:48.467 [INFO][6] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/profiles" error=ListOptions.meta.k8s.io "" is invalid: [resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden unless
2021-12-18 03:28:48.853 [INFO][6] watchercache.go 175: Full resync is required ListRoot="/calico/resources/v3/projectcalico.org/kubernetesendpointslices"
W1218 03:28:48.928928       6 warnings.go:70] discovery.k8s.io/v1beta1 EndpointSlice is deprecated in v1.21+, unavailable in v1.25+; use discovery.k8s.io/v1 EndpointSlice
W1218 03:28:48.975077       6 warnings.go:70] discovery.k8s.io/v1beta1 EndpointSlice is deprecated in v1.21+, unavailable in v1.25+; use discovery.k8s.io/v1 EndpointSlice
```

Please help, it's a pretty large cluster being offline now. Thanks!

## Expected Behavior

Keep connectivity after the upgrades

## Your Environment
* Calico version 1.21.1
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubeadm 1.22.5
* Operating System and version: ubuntu 20.04

Calico in IPIP mode, with typha

**Comment by caseydavenport** [comment]
@dimm0 if possible, please provide the full calico/typha and calico/node logs - those will help us diagnose a bit more. 

> 2021-12-18 03:28:48.467 [INFO][6] watchercache.go 188: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/profiles" error=ListOptions.meta.k8s.io "" is invalid: [resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden unless

This is most likely the problem. It seems to be coming from here: https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/internalversion/validation/validation.go#L33

That's old Kubernetes code, so I suspect this is a Calico issue of some sort. Calico sets that parameter only when revision is present, which should be the correct behavior to my knowledge: https://github.com/projectcalico/libcalico-go/blob/2f7d4c7d7a1fe75f183fc42c3cb987c3a2e33cb4/lib/backend/k8s/resources/list_pager.go#L43-L46

Something might be subtly wrong with that code, though. I will investigate. I haven't seen this error yet on any other Calico / Kubernetes clusters v1.22+ or otherwise.

**Comment by caseydavenport** [comment]
For reference, semantics on that field are defined here: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list 

We only ever set ResourceVersionMatchNotOlderThan, which should be valid for revision != ""

**Comment by caseydavenport** [comment]
Looking at your logs, it shows that only profiles, workloadendpoints, and kubernetesendpointslices are affected - if that's true (and not just a result of log clipping) that might be telling, as those are all "meta" resources backed by other Kubernetes resources rather than CRDs.

**Comment by caseydavenport** [comment]
> ListOptions.meta.k8s.io "" is invalid: 

This is an interesting error, since it appears to be attempting to log out the ListOptions but is showing an empty string. Wonder if that is indicating perhaps something wrong on the server side. If possible, it might be worth trying to see if this happened before or only after the upgrade of your Kubernetes API server.

**Comment by dimm0** [comment]
[Archive.zip](https://github.com/projectcalico/calico/files/7738788/Archive.zip)

logs for typha and node

**Comment by dimm0** [comment]
Before the k8s upgrade it worked fine for a day, I did not see any issues

**Comment by dimm0** [comment]
Should I try downgrading calico to a version with non optimized list watch?
Does calico support downgrades?

**Comment by dimm0** [comment]
Downgraded calico to 3.20.3

Typha can now connect to API, but connectivity is still broken

**Comment by caseydavenport** [comment]
At first I thought this might be a duplicate of https://github.com/projectcalico/calico/issues/5173, but it seems these are separate problems with similar symptoms.

**Comment by dimm0** [comment]
I ended up downgrading to 3.20 (this solved the problem with typha) and converting from route reflectors to full mesh (not sure still why route reflectors stopped working)

**Comment by caseydavenport** [comment]
> from route reflectors to full mesh (not sure still why route reflectors stopped working)

Hm, yeah this is curious. I wonder if this is two separate issues then.

**Comment by dimm0** [comment]
Duplicate of https://github.com/projectcalico/calico/issues/5173, most likely fixed. Will reopen if not
